docs/proposals/20220728-pod-probe-marker.md

```diff
@@ -0,0 +1,135 @@
+---
+title: PodProbeMarker
+authors:
+- "@zmberg"
+reviewers:
+- "@furykerry"
+- "@FillZpp"
+creation-date: 2022-08-09
+last-updated: 2021-08-09
+status: implementable
+---
+
+# Pod Probe Marker
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Title](#title)
+- [Table of Contents](#table-of-contents)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [API Definition](#api-definition)
+- [Implementation](#implementation)
+- [StaticIP Scheduler Practice](#staticip-scheduler-practice)
+
+## Motivation
+Kubernetes provides two probes by default for Pod lifecycle management:
+- **Readiness Probe** is used to determine whether the business container is ready to respond to requests, and if it fails, the Pod will be removed from the Service Endpoints.
+- **Liveness Probe** is used to determine the health status of the container, and if it fails, Kebelet will restart the Container.
+
+**So K8S on the provision of Probe capabilities are limited to specific semantics and behavior. In addition, there are some business applications that have the need to customize the semantics and behavior of the Probe**, such as:
+- GameServer defines Idle Probe to determine whether there is a game match for the current Pod, if there is no match, from cost optimization considerations can be given priority to offline this Pod
+- Operator defines Master-Slave Probe to determine the role of the current Pod (master or slave), and upgrade the Slave node in priority
+
+## Proposal
+OpenKruise provides the ability to customize Container Probe, Kruise Daemon executes customize Probe scripts and returns the results to the Pod yaml.
+
+### API Definition
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: PodProbeMarker
+metadata:
+  name: game-idle-probe
+spec:
+  targetRef:
+    apiVersion: apps.kruise.io/v1alpha1
+    # cloneset, deployment, statefulset etc.
+    kind: CloneSet
+    name: game-server
+  selector:
+    matchLabels:
+      app: game-server
+  containers:
+  - name: gameserver
+    probes
+    - Name: Idle
+      Exec: /home/game/idle.sh
+      # HttpGet: xxxx
+      # TcpSocket: xxxx
+    - Name: Master
+      Exec: /home/game/master.sh
+  markItems:
+    # When probe execution is successful
+  - probeName: Master
+    expectation: succeeded
+    labels:
```

Comment:
plz enumerate possible actions, i think the following action is possible
1. patch labels
2. patch annotations
3. patch readiness gate conditions